Cli migration helpers

smotornyuk opened the pull request:
Provide a unified workflow for:

* creating core migrations
* creating extension's migrations
* applying/reverting extension's migrations

There is one new command - `ckan generate migration -p PLUGIN -m Message` (create migration script and initialize migration repository inside extension, if used for first time) and new option `-p/--plugin` to `db upgrade`, `db downgrade` and `db version`, that uses plugin's migration scripts instead of core ones, when provided

wardi commented:
"generate" seems like a subcommand of "migrate". what do you think of changing the order to `ckan migration generate ...`

amercader commented:
@smotornyuk with latest master, when trying:

     ckan generate migration -m "Test migration"

I get:

```
Traceback (most recent call last):
  File "/home/adria/dev/pyenvs/ckan/bin/ckan", line 11, in <module>
    load_entry_point('ckan', 'console_scripts', 'ckan')()
  File "/home/adria/dev/pyenvs/ckan/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/adria/dev/pyenvs/ckan/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/adria/dev/pyenvs/ckan/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/adria/dev/pyenvs/ckan/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/adria/dev/pyenvs/ckan/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/adria/dev/pyenvs/ckan/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/adria/dev/pyenvs/ckan/src/ckan/ckan/cli/generate.py", line 152, in migration
    str(ckan.model.repo.metadata.bind.url))
AttributeError: 'NoneType' object has no attribute 'url'

```

Additionally if I try the `-p` option I get:

```
ckan generate migration -p dcat -m "Test migration"
Plugin 'dcat' cannot be loaded.
Aborted!
```

`ckanext-dcat` is loaded in the venv and `dcat` is loaded in `ckan.plugins`

smotornyuk commented:
Updated. Recently we have disabled loading of the config file for `generate` (because of `generate config`) command. I've updated that logic so that the config file won't be loaded for `generate` command only when it's not available.

markstuart commented on ckan/cli/generate.py:
```diff
@@ -2,35 +2,51 @@
+        tk.error_shout(u"`cookiecutter` library is missing from import path.")
+        tk.error_shout(u"Make sure you have dev-dependencies installed:")
+        tk.error_shout(u"\tpip install -r dev-requirements.txt")
+        raise click.Abort()
```
Ah, this has been moved out of the decorator here as well as in my PR #5381 @kowh-ai 👍
